system/lib/fetch/asmfs.cpp

```diff
@@ -1879,78 +1879,61 @@ long __syscall197(int which, ...) // SYS_fstat64
 // TODO: syscall207: fchown32
 // TODO: syscall212: chown32
 
-long __syscall220(int which, ...) // getdents64 (get directory entries 64-bit)
```

Comment:
Can you use that same trick and just override readdir here?   Thus avoiding the extra dependencies and JS hop?

Before this change the native version of `__syscall220` would end up meaning it was never imported.. after this change the native readdir here would mean `readdir` was never imported?  Would that work?

Reply:
Hmm not sure I quite understand. What exactly are you suggesting? Are you saying implement opendir()/readdir() in this file directly?

Reply:
Yes, well its just readdir right?  It looks like the point of this function is to override readdir()?   So why not just override it here, rather than going out to JS and the calling back into this new function called `emscripten_asmfs_readdir` which didn't exist before?

With this override the linker should then choose this version rather than the JS version, no?

Reply:
Yeah you should be right. I originally opted to not go in that direction to reduce code duplication but if I move the opendir/readdir code to library_fs, then there will be code duplication anyway. If I put the main code in library_fs and the user sets FILESYSTEM=0, do you expect that to be a compile-time error or a runtime-error?

Reply:
Perhaps I'm misunderstanding something.    My understanding is that this function here is designed to replace the normal "readdir" when the users whats to use asmfs.

In that case why not just have this version be called "readdir" and for those users they will never import a "readdir" from the output because they are effectively providing their own.  How does that make for code duplication?

Sorry for all the feedback, I'll try to understand this code a little better tomorrow :)

Reply:
In the ASMFS case, in the current version of the diff, the path from a normal call to readdir() is this:

readdir() -> library.js -> library_asm.js -> asmfs.cpp

This allows me to use all the bookkeeping logic in library.js. What your suggesting would change that to this:

readdir() -> asmfs.cpp

But I would have to reimplement the `DIR*` bookkeeping logic in asmfs.cpp, which is fine, but then we'd have two readdir() implementations, one in C and one in JS. I think what you're suggesting is more in line with how asmfs.cpp was meant to be, so I will change it to that.